tests/unit/test_user.py

```diff
@@ -65,3 +66,24 @@ def test_multiple_users_can_be_created():
             return create_unauthorized_user()
 
         assert test_multiple_users_can_be_created()
+
+
+@pytest.mark.usefixtures("empty_test_db")
+class TestOverwrittenAndInheritedMethods:
+    def test_user_save_to_db(self, user_attributes):
+        user = UserModel(**user_attributes())
+        user.save_to_db()
+        lookedup = UserModel.find_by_id(user.id)
+        assert lookedup == user
+
+    def test_update_class_method(self, create_join_staff, faker):
+        user = create_join_staff()
+        email = faker.unique.email()
+        UserModel.update(UserSchema, user.id, {"email": email})
+        lookedup = UserModel.find_by_id(user.id)
+        assert lookedup.email == email
+
+    def test_create_class_method(self, user_attributes):
+        user = UserModel.create(UserSchema, user_attributes(role=RoleEnum.STAFF.value))
+        lookedup = UserModel.find_by_id(user.id)
+        assert lookedup == user
```

Comment:
At first glance these look great, but when you dig deeper it turns out these aren't actually testing that the overridden `save_to_db` method from `UserModel` is the one that is executed.  These tests still pass even if that method is deleted.

The only difference between the `UserModel`'s method and the one in the `BaseModel` is this conditional:
https://github.com/codeforpdx/dwellinglybackend/blob/53c761cef25ae14efb7acefad824d216da7c16e9/models/user.py#L134-L136

However, we _also_ can't just assert that the user's `hash_digest` exists, because it turns out that `hash_digest` is created in two places:
- in `UserModel`'s `save_to_db` method as mentioned above;
- and also in the `UserModel` init method:
  https://github.com/codeforpdx/dwellinglybackend/blob/53c761cef25ae14efb7acefad824d216da7c16e9/models/user.py#L48-L51

This means that the `hash_digest` is created on instantiation so long as a password is in the attributes, and it is then _changed_ with each `save_to_db` (or at least it should be if the overriding method is properly called).

To demonstrate this, run the tests by checking if the `hash_digests` are the same between `user` (pre-update/save) and `lookedup`: 

```python
class TestOverwrittenAndInheritedMethods:
    def test_user_save_to_db(self, user_attributes):
        user = UserModel(**user_attributes())

        hashed = user.hash_digest

        user.save_to_db()
        lookedup = UserModel.find_by_id(user.id)
        
        assert lookedup == user
        assert lookedup.hash_digest == hashed

    def test_update_class_method(self, create_join_staff, faker):
        user = create_join_staff()
        email = faker.unique.email()

        hashed = user.hash_digest

        UserModel.update(UserSchema, user.id, {"email": email})
        lookedup = UserModel.find_by_id(user.id)

        assert lookedup.email == email
        assert lookedup.hash_digest == hashed

    def test_create_class_method(self, user_attributes):
        user = UserModel.create(UserSchema, user_attributes(role=RoleEnum.STAFF.value))

        hashed = user.hash_digest

        lookedup = UserModel.find_by_id(user.id)
        assert lookedup == user
        assert lookedup.hash_digest == hashed
```
The first 2 tests will fail because the `hash_digest` is changed by the `save_to_db` and `update` methods. The third test will pass because the `hash_digest` is only created once &mdash; through the `create` method.

Reply:
Now, it's a good sign they fail, because that means that `UserModel`'s method is indeed the one being called each time.

However, it begs the question: should we really be creating a new `hash_digest` every time we save or update the user, _even when_ the `hash_digest` already exists? 

It might make more sense to only create a `hash_digest` when the user is created and when their password is updated...but that might also be more work than it's worth since creating a new `hash_digest` on every save doesn't seem to be hurting anything.  Interested to hear what you and @NickSchimek think!

Reply:
I haven't read what you wrote, but I saw this PR come up and was meaning to comment on it. At first glance everything looked good, and does what the description said. However, I wrote that description for me as I had planned to do this. And when I unassigned myself I never updated the description. This is all my fault for not updating the description for what kind of tests need to be written. I apologize. The description I put in was for reasons we are not removing the `save_to_db` method, and not what kind of tests to write. I will take a look at this tomorrow, but I think what we do need are some base interface tests.

Reply:
hey @aedwardg this is a good write up and your spot on how we want to check the conditional in the save_to_db method. 

Good catch on the hashing of the password twice, although I think this is there to be backward compatible while we do all the refactoring. Once all the places have been updated to use the `create` method then I believe we can remove the hashing of the password in the User init method. I'm not 100% sure about the reason, but I remember there was a reason why I did this, and it is only done on create, not any other time. Maybe @cecillesalazar remembers. In your test you used the User init method to create a user `UserModel(**user_attributes())` <- We need to move away from this, because it bypasses validations, and all the checks that we have put in place to ensure that a user is properly created. So yes, I think currently it hashes a password on init, and again on the save_to_db method, but maybe the one in the `init` can be removed once the refactoring for creating all the User's have been updated to use the `create` method. 

I also noticed the User factories are not using the `create` method, which they should be.

The second test is failing, because the db config is failing here. I'm not sure why 🤦 ; otherwise the password are not rehashed when the user does not update the password. Actually I think I might know why the db config is failing. <- This is going to be a high priority fix. It is real simple to fix, but I'd like to tighten up if we can. I'd be happy to go over the architecture put in place, and come up with a way to improve it with whoever wants to take this on.

Good analysis @aedwardg - Lets get the second issue fixed yesterday, and the first issue will hopefully take care of itself in time.